Added example for SWR Request De-duplication feature (#519)
vercel/swr: README.md

- [Suspense Mode](#suspense-mode)
- [Error Retries](#error-retries)
- [Prefetching Data](#prefetching-data)
- [Request Deduplication](#request-deduplication)

### Global Configuration

The context `SWRConfig` can provide global configurations (`options`) for all SWR hooks.

In this example, all SWRs will use the same fetcher provided to load JSON data, and refresh every 3 seconds by default:

```js
import useSWR, { SWRConfig } from 'swr'

function Dashboard() {
  const { data: events } = useSWR('/api/events')
  const { data: projects } = useSWR('/api/projects')
  const { data: user } = useSWR('/api/user', { refreshInterval: 0 }) // don't refresh
  // ...
}

function App() {
  return (
    <SWRConfig
      value={{
        refreshInterval: 3000,

>         fetcher: (...args) => fetch(...args).then((res) => res.json()),

Reviewer: Thank you for doing code linting! However we don't have a preferred style in our contribution guide, so I'd keep things like these unedited. Also it is for the better consistency with the [documentation site](https://swr.vercel.app).

Author: @shuding agreed! I have reverted the changes that affected the code examples. Minor spaces and indentations are kept.

And use it when you need to preload the **resources** (for example when [hovering](https://github.com/GoogleChromeLabs/quicklink) [a](https://github.com/guess-js/guess) [link](https://instant.page)).  
Together with techniques like [page prefetching](https://nextjs.org/docs#prefetching-pages) in Next.js, you will be able to load both next page and data instantly.

### Request Deduplication

> SWR provides request deduplication by default. So, if you call the hook with same key multiple times, it will return the cached value ie. requests will not be duplicated.

Reviewer 2: ```suggestion
SWR deduplicates requests by default. If you call the hook with the same key multiple times, only one request is made. Duplicated calls will receive a value from cache.
```

Only one request will be made and once the promise is fulfilled, it will update both components.

> Here, 'api/user' is the duplicate key

Reviewer 2: ```suggestion
Here, the 'api/user' key is used in two requests:
```

```js
import useSWR from 'swr'

function UserProfileName() {
  const { data, error } = useSWR('/api/user', fetcher)

  if (error) return <div>failed to load</div>
  if (!data) return <div>loading...</div>
  return <p>Name: {data.name}!</p>
}

function UserProfileAvatar() {
  const { data, error } = useSWR('/api/user', fetcher)

  if (error) return <div>failed to load</div>
  if (!data) return <div>loading...</div>
  return <img src={data.avatarUrl} alt="Profile image" />
}

export default function App() {
  return (
    <div>
      <UserAvatar />
      <UserProfile />
    </div>
  )
}

> ```

Reviewer 2: At this point I think we should explain the `dedupingInterval` option, so you can customize how the deduping works.

Author: Thanks for the suggestion. Let me add that

Author: Done. @pacocoursey  Please check

And use it when you need to preload the **resources** (for example when [hovering](https://github.com/GoogleChromeLabs/quicklink) [a](https://github.com/guess-js/guess) [link](https://instant.page)).  
Together with techniques like [page prefetching](https://nextjs.org/docs#prefetching-pages) in Next.js, you will be able to load both next page and data instantly.

### Request Deduplication

SWR deduplicates requests by default. If you call the hook with the same key multiple times, only one request is made. Duplicated calls will receive a value from cache.

> Only one request will be made and once the promise is fulfilled, it will update both components.
> Here, 'api/user' is the duplicate key

Reviewer 2: ```suggestion
Here, the 'api/user' key is used in two requests:
```

Author: Yes, sounds better.

```js
import useSWR from 'swr'

function UserProfileName() {
  const { data, error } = useSWR('/api/user', fetcher)

  if (error) return <div>failed to load</div>
  if (!data) return <div>loading...</div>
  return <p>Name: {data.name}!</p>
}

function UserProfileAvatar() {
  const { data, error } = useSWR('/api/user', fetcher)

  if (error) return <div>failed to load</div>
  if (!data) return <div>loading...</div>
  return <img src={data.avatarUrl} alt="Profile image" />
}

export default function App() {
  return (
    <div>
      <UserAvatar />
      <UserProfile />
    </div>
  )
}
```

> The default interval to dedupe a request is 2 seconds. This can be changed by setting the 'dedupingInterval' option. For example:

Reviewer 2: ```suggestion
By default, requests made within 2 seconds are deduped. This can be changed by setting the `dedupingInterval` option:
```

```js
const { data, error } = useSWR('/api/user', fetcher, { dedupingInterval: 1000 })
```

> This will duplicate the request at interval of 1 second

Reviewer 2: ```suggestion
This will deduplicate requests at an interval of 1 second.
```
